File: android/app/src/main/java/io/mimiri/app/MimiriUpdatePlugin.java
package io.mimiri.app;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.SharedPreferences;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.Base64;
import java.util.Objects;
import java.util.zip.GZIPInputStream;

@CapacitorPlugin(name = "MimiriUpdate")
public class MimiriUpdatePlugin extends Plugin {
  private boolean _initialized = false;
  private File _bundlesPath;
  private File _configPath;
  private JSObject _config;

  private void init() {
    if (!_initialized) {
      _initialized = true;
      try {
        _bundlesPath = new File(getContext().getFilesDir(), "bundles");
        _configPath = new File(_bundlesPath, "config.json");
        if (_bundlesPath.exists() || _bundlesPath.mkdirs()) {
          if (_configPath.exists()) {
            _config = FileUtil.readJson(_configPath);
          }
        }
        if (_config == null) {
          _config = new JSObject();
          _config.put("activeVersion", "base");
        }
      } catch (Exception e) {
        _bundlesPath = new File("");
        _configPath = new File("");
        System.out.println(e.toString());
      }

    }
  }

  @PluginMethod()
  public void getInstalledVersions(PluginCall call) {
    init();
    JSObject result = new JSObject();
    JSArray bundles = new JSArray();

    JSObject baseBundle = new JSObject();
    bundles.put(baseBundle);
    baseBundle.put("version", VersionInfo.BaseVersion);
    baseBundle.put("hostVersion", VersionInfo.HostVersion);
    baseBundle.put("base", true);
    baseBundle.put("description", "base");
    baseBundle.put("releaseDate", VersionInfo.ReleaseDate);
    baseBundle.put("active", "base".equals(_config.getString("activeVersion")));
    baseBundle.put("previous", "base".equals(_config.getString("previousActiveVersion")));

    for (final File file : Objects.requireNonNull(_bundlesPath.listFiles())) {
      File infoPath =  new File(file, "info.json");
      if (infoPath.exists()) {
        JSObject info = FileUtil.readJson(infoPath);
        if (info != null) {
          String version = info.getString("version", "");
          info.put("hostVersion", VersionInfo.HostVersion);
          info.put("base", false);
          assert version != null;
          info.put("active", version.equals(_config.getString("activeVersion")));
          info.put("previous", version.equals(_config.getString("previousActiveVersion")));
          bundles.put(info);
        }
      }
    }
    result.put("bundles", bundles);
    call.resolve(result);
  }

  private void saveFileRecursive(File dir, JSONObject item) throws JSONException, IOException {
    init();
    File path = new File(dir, item.getString("name"));
    if (item.has("files")) {
      if (path.exists() || path.mkdirs()) {
        JSONArray files = item.getJSONArray("files");
        for (int i = 0; i < files.length(); i++) {
          saveFileRecursive(path, files.getJSONObject(i));
        }
      }
    } else {
      String content = item.getString("content");
      try (GZIPInputStream zipStream = new GZIPInputStream(new ByteArrayInputStream(Base64.getDecoder().decode(content)))) {
        Files.copy(zipStream, path.toPath(), StandardCopyOption.REPLACE_EXISTING);
      }
    }
  }

  @PluginMethod()
  public void save(PluginCall call) {
    init();
    try {
      String version = call.getString("version");
      JSObject bundle = call.getObject("bundle");
      if (version != null && bundle != null) {
        File bundlePath = new File(_bundlesPath, version);
        System.out.println("save " + bundlePath);
        if (bundlePath.exists() || bundlePath.mkdirs()) {
          JSONArray files = bundle.getJSONArray("files");
          for (int i = 0; i < files.length(); i++) {
            saveFileRecursive(bundlePath, files.getJSONObject(i));
          }
        }
        bundle.remove("files");
        bundle.remove("signatures");
        File infoPath = new File(bundlePath, "info.json");
        FileUtil.writeJson(infoPath, bundle);
      }
    } catch (Exception e) {
      System.out.println(e.toString());
    }
    call.resolve();
  }

  @SuppressLint("ApplySharedPref")
  @PluginMethod()
  public void use(PluginCall call) {
    init();
    String version = call.getString("version");
    if (version != null && !version.equals(_config.getString("activeVersion"))) {
      _config.put("previousActiveVersion", _config.getString("activeVersion"));
      _config.put("activeVersion", version);
      FileUtil.writeJson(_configPath, _config);
      String serverBasePath = "";
      if (!version.equals("base")) {
        serverBasePath = new File(_bundlesPath, version).toString();
      }
      System.out.println("use " + serverBasePath);
      SharedPreferences.Editor webViewSettingsEditor = getContext().getSharedPreferences("CapWebViewSettings", Activity.MODE_PRIVATE).edit();
      webViewSettingsEditor.putString("serverBasePath", serverBasePath);
      webViewSettingsEditor.commit();
      getBridge().setServerBasePath(serverBasePath);
      getBridge().reload();
    }
    call.resolve();
  }

  @PluginMethod()
  public void delete(PluginCall call) {
    init();
    String version = call.getString("version");
    if (version != null && !version.equals(_config.getString("activeVersion")) && !version.equals("base")) {
      File versionPath = new File(_bundlesPath, version);
      if (versionPath.exists()) {
        FileUtil.deleteDir(versionPath);
      }
    }
    call.resolve();
  }

  @PluginMethod()
  public void good(PluginCall call) {
    init();
    String version = call.getString("version");
    if (version != null && version.equals(_config.getString("activeVersion")) && !version.equals("base")) {
      File bundlePath = new File(_bundlesPath, version);
      File infoPath = new File(bundlePath, "info.json");
      if (infoPath.exists()) {
        JSObject info = FileUtil.readJson(infoPath);
        if (info != null && Boolean.FALSE.equals(info.getBoolean("good", false))) {
          info.put("good", true);
          FileUtil.writeJson(infoPath, info);
        }
      }
    }
    call.resolve();
  }

}

File: ios/App/App/MimiriUpdatePlugin.swift
import Foundation
import Capacitor

@objc(MimiriUpdatePlugin)
public class MimiriUpdatePlugin: CAPPlugin, CAPBridgedPlugin {
  private var _bundlesPath = URL(fileURLWithPath: "")
	private var _configPath = URL(fileURLWithPath: "")
	private var _initialized = false
  private var _config: [String: Any] = [:]
  public let identifier = "MimiriUpdatePlugin"
  public let jsName = "MimiriUpdate"
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "getInstalledVersions", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "save", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "use", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "delete", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "good", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "activate", returnType: CAPPluginReturnPromise)
	]

  private func ensureConfig() {
    if (!_initialized) {
      _initialized = true;
      do {
        if let dir = FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).first {
          _bundlesPath = dir.appendingPathComponent("NoCloud/ionic_built_snapshots/")
          if (!FileManager.default.fileExists(atPath: _bundlesPath.path)) {
            try FileManager.default.createDirectory(at: _bundlesPath, withIntermediateDirectories: true)
          }
          _configPath = _bundlesPath.appendingPathComponent("config.json")
          if (FileManager.default.fileExists(atPath: _configPath.path)) {
            let json = try Data(contentsOf: _configPath)
            _config = try JSONSerialization.jsonObject(with: json) as? [String: Any] ?? [ "activeVersion": "base" ]
          } else {
            _config = [ "activeVersion": "base" ]
          }
        }

      } catch {
        print(error.localizedDescription)
      }
    }
  }

  @objc func getInstalledVersions(_ call: CAPPluginCall) {
		ensureConfig()
    var bundles = [[String: Any]]()
    do {
      bundles.append([
        "version": VersionInfo.BaseVersion,
        "hostVersion": VersionInfo.HostVersion,
				"base": true,
				"description": "base",
				"releaseDate": VersionInfo.ReleaseDate,
        "active": _config["activeVersion"] as? String == "base",
        "previous": _config["previousActiveVersion"] as? String == "base",
      ])
      for item in try FileManager.default.contentsOfDirectory(at: _bundlesPath, includingPropertiesForKeys: nil, options: []) {
        let infoPath = item.appendingPathComponent("info.json")
        if (FileManager.default.fileExists(atPath: infoPath.path)) {
          let json = try Data(contentsOf: infoPath)
          var info = try JSONSerialization.jsonObject(with: json) as? [String: Any] ?? [:]
          info["hostVersion"] = VersionInfo.HostVersion
          info["base"] = false
          info["active"] = _config["activeVersion"] as? String == info["version"] as? String
          info["previous"] = _config["previousActiveVersion"]as? String == info["version"] as? String
          bundles.append(info)
        }
      }

    } catch {
      print(error.localizedDescription)
    }
    call.resolve(["bundles": bundles])
  }

  func saveFileRecursive(dir: URL, item: JSObject) throws {
    let name = item["name"] as? String
    let files = item["files"] as? JSArray
    let path = dir.appendingPathComponent(name!)
    if (files != nil) {
      if (!FileManager.default.fileExists(atPath: path.path)) {
        try FileManager.default.createDirectory(at: path, withIntermediateDirectories: true)
      }
      for (_, file) in files!.enumerated() {
        try saveFileRecursive(dir: path, item: file as! JSObject)
      }
    } else {
      if let content = item["content"] as? String {
        if let data = Data(base64Encoded: content.data(using: .utf8)!) {
          let inflated = try data.gunzipped()
          try inflated.write(to: path)
        }
      }
    }
  }

  @objc func save(_ call: CAPPluginCall) {
    ensureConfig();
    do {
      let version = call.getString("version")
      let files = call.getObject("bundle")!["files"] as? JSArray
      if (files != nil) {
        let bundlePath = _bundlesPath.appendingPathComponent(version!)
        if (!FileManager.default.fileExists(atPath: bundlePath.path)) {
          try FileManager.default.createDirectory(at: bundlePath, withIntermediateDirectories: true)
        }
        for (_, file) in files!.enumerated() {
          try saveFileRecursive(dir: _bundlesPath.appendingPathComponent(version!), item: file as! JSObject)
        }
      }
      if var bundle = call.getObject("bundle") {
        let infoPath = _bundlesPath.appendingPathComponent(version!).appendingPathComponent("info.json")
        bundle.removeValue(forKey: "files")
        bundle.removeValue(forKey: "signatures")
        let json = try JSONSerialization.data(withJSONObject: bundle, options: [.withoutEscapingSlashes, .prettyPrinted])
        try json.write(to: infoPath)
      }
    } catch {
      print(error.localizedDescription)
    }

    //print(call.getObject("bundle"))
    call.resolve([:])
  }

  @objc func use(_ call: CAPPluginCall) {
    ensureConfig();
    do {
      let version = call.getString("version")
      let noActivate = call.getBool("noActivate") ?? false
      let activeVersion = _config["activeVersion"] as? String
      if (version != nil && activeVersion != nil && version != activeVersion) {
        _config["previousActiveVersion"] = activeVersion
        _config["activeVersion"] = version
        let json = try JSONSerialization.data(withJSONObject: _config, options: [.withoutEscapingSlashes, .prettyPrinted])
        try json.write(to: _configPath)
        if (version != "base") {
          KeyValueStore.standard["serverBasePath"] = version
        }
      }
			if (version != nil && !noActivate) {
				(self.bridge!.viewController! as? CAPBridgeViewController)!.setServerBasePath(path: _bundlesPath.appendingPathComponent(version!).path)
			}
    } catch {
      print(error.localizedDescription)
    }
    call.resolve([:])
  }

  @objc func activate(_ call: CAPPluginCall) {
    ensureConfig();
    let version = _config["activeVersion"] as? String
    if (version != nil && version != "base") {
      KeyValueStore.standard["serverBasePath"] = version
      (self.bridge!.viewController! as? CAPBridgeViewController)!.setServerBasePath(path: _bundlesPath.appendingPathComponent(version!).path)
    }
    call.resolve([:])
  }

  @objc func delete(_ call: CAPPluginCall) {
    ensureConfig();
    do {
      let version = call.getString("version")
      let activeVersion = _config["activeVersion"] as? String
      if (version != nil && activeVersion != nil && version != activeVersion) {
        try FileManager.default.removeItem(atPath: _bundlesPath.appendingPathComponent(version!).path)
      }
    } catch {
      print(error.localizedDescription)
    }
    call.resolve([:])
  }

  @objc func good(_ call: CAPPluginCall) {
    do {
      let version = call.getString("version")
      let activeVersion = _config["activeVersion"] as? String
      if (version != nil && activeVersion != nil && version == activeVersion) {
      	let infoPath = _bundlesPath.appendingPathComponent(version!).appendingPathComponent("info.json")
          let json = try Data(contentsOf: infoPath)
          var info = try JSONSerialization.jsonObject(with: json) as? [String: Any]
					if (info != nil) {
            info!["good"] = true
            let json = try JSONSerialization.data(withJSONObject: info!, options: [.withoutEscapingSlashes, .prettyPrinted])
            try json.write(to: infoPath)
					}
			}
    } catch {
      print(error.localizedDescription)
    }
    call.resolve([:])
  }
}

